### PR TITLE
TIP-674: Dissociate the Product Value model from its EAV implementation

### PR DESCRIPTION
**Pull request by skeleton**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

- [x] create a ProductValue object per type of attribute we manipulate
  - [x] MetricProductValue
  - [x] MediaProductValue
  - [x] OptionProductValue
  - [x] OptionsProductValue
  - [x] PriceCollectionProductValue
  - [x] DateProductValue
- [x] Remove useless ProductValueAccessorsChecker and ProductValueRelationShipChecker for reference data
- [x] Work with array instead of ArrayCollection in OptionsProductValueFactory & ReferenceDataCollectionProductValueFactory

- [ ] Move all classes related to ProductValue from `Model` folder to `ProductValue` : Will be done in an other PR to avoid to decrease readability of this one

[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md)

**Description (for Contributor and Core Developer)**

[//]: <> (What does this Pull Request do? reference the related issue?)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added Behats                      | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Component/Catalog/ProductValue/ProductValue.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ *   - pim_catalog_text
+ *   - pim_catalog_textarea
+ *   - pim_catalog_boolean
+ *   - pim_catalog_number
+ *
+ * @author    Marie Bochu <marie.bochu@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductValue extends AbstractProductValue implements ProductValueInterface
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/ProductValue/ProductValue.php]: ScalarProductValue? That's what we used for the factory.

File: src/Acme/Bundle/AppBundle/Model/FabricsProductValue.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        $fabrics = [];
+        foreach ($this->data as $fabric) {
+            $fabrics[] .= $fabric->getCode();
```
Comment (anaelChardan) [inline src/Acme/Bundle/AppBundle/Model/FabricsProductValue.php]: Why do you use the 'dot' ? :)

**Review by damien-carcel — APPROVED** [review]
So beautiful to see I could cry!

File: src/Pim/Component/Catalog/spec/ProductValue/MediaProductValueSpec.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    function let(AttributeInterface $attribute, FileInfoInterface $fileInfo)
+    {
+        $attribute->isScopable()->willReturn(true);
+        $attribute->isLocalizable()->willReturn(true);
+        $this->beConstructedWith($attribute, 'ecommerce', 'en_US', $fileInfo);
+    }
+
+    function it_returns_data($fileInfo)
+    {
+        $this->getData()->shouldBeAnInstanceOf(FileInfoInterface::class);
```
Comment (anaelChardan) [inline src/Pim/Component/Catalog/spec/ProductValue/MediaProductValueSpec.php:20]: As it is an Interface, I think that shouldImplement would be better :)

**Review by SamirBoulil — APPROVED** [review]
👍  ⚡️  🔥  💯  💃  🗡  LGTM 👍  ⚡️  🔥  💯  💃  🗡

**Review by anaelChardan — APPROVED** [review]

File: src/Pim/Component/Catalog/ProductValue/DateProductValue.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+use Pim\Component\Catalog\Model\AttributeInterface;
+
+/**
+ * Product value for "pim_catalog_date" attribute types
+ *
+ * @author    Marie Bochu <marie.bochu@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class DateProductValue extends AbstractProductValue implements MediaProductValueInterface
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/ProductValue/DateProductValue.php]: wrong implement ;)

File: src/Pim/Component/Catalog/ProductValue/OptionProductValue.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        return $this->data;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        return $this->data;
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/ProductValue/OptionProductValue.php]: ->getCode()

File: src/Pim/Component/Catalog/ProductValue/OptionsProductValue.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+        return $options;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        $options = [];
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/ProductValue/OptionsProductValue.php]: you can use getOptionCodes maybe

File: src/Pim/Component/Catalog/ProductValue/ProductValue.php (hunk 1 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/ProductValue/ProductValue.php]: why not
seems a good idea

File: src/Pim/Component/Catalog/ProductValue/ProductValue.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        return $this->data;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        return $this->data;
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/ProductValue/ProductValue.php]: (string) $this->data

File: src/Pim/Component/Catalog/Validator/Constraints/UniqueVariantAxisValidator.php [hunk 8]
```diff
@@ -72,7 +72,7 @@ protected function validateVariantGroup(GroupInterface $variantGroup, Constraint
                     $this->addEmptyAxisViolation(
                         $constraint,
                         $variantGroup->getLabel(),
-                        $product->getIdentifier()->getVarchar(),
+                        $product->getIdentifier(),
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Validator/Constraints/UniqueVariantAxisValidator.php:75]: :+1: oups, we missed this one :)

File: src/Pim/Component/ReferenceData/spec/Updater/Copier/ReferenceDataCollectionAttributeCopierSpec.php [hunk 9]
```diff
@@ -55,8 +57,8 @@ function it_copies_reference_data_value_to_a_product_value(
         AttributeInterface $toAttribute,
         ProductInterface $product1,
         ProductInterface $product2,
-        CustomProductValueBis $fromProductValue,
-        CustomProductValueBis $toProductValue,
+        ProductValueInterface $fromProductValue,
+        ProductValueInterface$toProductValue,
```
Comment (jjanvier) [inline src/Pim/Component/ReferenceData/spec/Updater/Copier/ReferenceDataCollectionAttributeCopierSpec.php]: space

**Review by jjanvier — COMMENTED** [review]
I would introduce a ReferenceDataProductValue (simple and multi), so that it works like the options

or at least an interface

even if the interface extends ProductValueInterface without adding new methods, as Reference Data are important enough for us, it would make sense imo

File: src/Pim/Component/ReferenceData/ProductValue/ReferenceDataCollectionProductValue.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    public function getData()
+    {
+        return $this->data;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOptionCodes()
```
Comment (jjanvier) [inline src/Pim/Component/ReferenceData/ProductValue/ReferenceDataCollectionProductValue.php]: not options ;)

File: src/Acme/Bundle/AppBundle/Model/ColorProductValue.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+use Pim\Component\Catalog\Model\AttributeInterface;
+
+/**
+ * Color product value for "color" reference data. It's a new many to one relationship
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ColorProductValue extends AbstractProductValue
```
Comment (jjanvier) [inline src/Acme/Bundle/AppBundle/Model/ColorProductValue.php]: do we need those values with the referencedata value you introduced?

seems no, which is even more awesome :)

**Review by jjanvier — APPROVED** [review]
:+1: :+1: :+1: :+1: :+1: :+1: :+1: 
GTM once fixed 

:tada:

**Comment by jjanvier** [comment]
missing a method apparently   
  [Symfony\Component\Debug\Exception\UndefinedMethodException]                                                               
  Attempted to call an undefined method named "getOption" of class "Pim\Component\Catalog\ProductValue\OptionProductValue". 

or we should replace the call to getOPtion

I had this error when I tried to install the PIM

**Comment by skeleton** [comment]
I fixed the error. catalog icecat is now installable on CE

**Pull request merged**